### PR TITLE
Support aggregation pipeline for Tigris

### LINKED ISSUE #1891: Support aggregation pipelines for Tigris
### What should be done?

Currently, the `aggregate` command is supported only by `pg` handler. We should port it to `tigris` handler.

Only manual testing is in the scope; the rest is in #1410.

### Where?

Port https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/pg/msg_aggregate.go
to https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/tigris/msg_aggregate.go

Do not try to DRY it; copy and adapt instead.

### Definition of Done

- basic `aggregate` command works for Tigris;
- manual testing performed;
- spot refactorings done.

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #1891.

## Readiness checklist

* [ ] I added/updated unit tests.
* [ ] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [x] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

**Comment by chilagrow** [comment]
Manual testing done 
```
test> db.test.insertMany([{v: 1},{v: 2}, {v: 3}])
{                                                   
  acknowledged: true,                               
  insertedIds: {                                    
    '0': ObjectId("63fecd706a4273f5e64d7b32"),
    '1': ObjectId("63fecd706a4273f5e64d7b33"),
    '2': ObjectId("63fecd706a4273f5e64d7b34")
  }                                                 
}  
test> db.test.aggregate([{$match:{v: 2}}])
[ { _id: ObjectId("63fecd706a4273f5e64d7b33"), v: 2 } ]
```

**Review by AlekSi — DISMISSED** [review]

**Review by w84thesun — DISMISSED** [review]

File: internal/handlers/tigris/msg_aggregate.go [hunk 1]
```diff
@@ -19,36 +19,134 @@ import (
...
+	// TODO pushdown `$match` https://github.com/FerretDB/FerretDB/issues/1894
+
+	var docs []*types.Document
+
+	iter, err := dbPool.QueryDocuments(ctx, &qp)
+	if err != nil {
+		return nil, err
+	}
+
+	defer iter.Close()
```
Comment (noisersup) [inline internal/handlers/tigris/msg_aggregate.go:120]: Are there any potential issues that would result from calling `Close` function twice? `iterator.Values()` will call `Close` at the end too.
If we left `Close` here just to be sure, and it doesn't affect anything then it's ok I guess!

**Review by noisersup — CHANGES_REQUESTED** [review]
Some notes from me, otherwise everything looks great! :+1:

File: internal/handlers/tigris/msg_aggregate.go [hunk 2]
```diff
@@ -19,36 +19,134 @@ import (
...
+	stages := make([]aggregations.Stage, len(stagesDocs))
+
+	for i, d := range stagesDocs {
+		d, ok := d.(*types.Document)
+		if !ok {
+			return nil, commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrBadValue,
+				"Each element of the 'pipeline' array must be an object",
+				document.Command(),
+			)
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/tigris/msg_aggregate.go:99]: Yes it _was_ changed. Please see the file above.

File: internal/handlers/tigris/msg_aggregate.go (hunk 1 above)
Comment (chilagrow) [reply internal/handlers/tigris/msg_aggregate.go:120]: `Close` call is idempotent, see the documentation :hugs: 

> Close must be concurrency-safe and may be called multiple times.
All calls after the first should have no observable effect.

File: internal/handlers/tigris/msg_aggregate.go [hunk 3]
```diff
@@ -19,36 +19,134 @@ import (
...
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/1891
+	common.Ignored(document, h.L, "allowDiskUse", "maxTimeMS", "collation", "comment")
+
+	var qp tigrisdb.QueryParams
+
+	qp.DB, err = common.GetRequiredParam[string](document, "$db")
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/tigris/msg_aggregate.go]: Let's do that.

File: internal/handlers/tigris/msg_aggregate.go (hunk 1 above)
Comment (AlekSi) [reply internal/handlers/tigris/msg_aggregate.go:119]: If it is not idempotent, that's a bug we should fix. Future changes for cursors and aggregation pipelines will require Close to be safe to be called anytime.

**Review by AlekSi — APPROVED** [review]

**Review by rumyantseva — APPROVED** [review]
Looks consistent to me!

**Review by w84thesun — APPROVED** [review]
👍

File: internal/handlers/tigris/msg_aggregate.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/handlers/tigris/msg_aggregate.go:98]: Ah, I see. I had too many windows with different versions of code open while reviewing stuff :laughing:

**Review by noisersup — APPROVED** [review]
:shipit:

**Pull request merged**